### PR TITLE
Add a theme collection component for TS and support for Premium and Partner collections

### PR DESCRIPTION
**Pull request by BogdanUngureanu**
This PR introduces a new ThemeCollection component that will group various types of themes (e.g. Premium, Partner). 

The implementation is currently behind a feature flag since the implementation is not complete: we have to properly style it and add an additional layout component that will display collections and banners, for example. 

Related to https://github.com/Automattic/dotcom-forge/issues/3959, https://github.com/Automattic/dotcom-forge/issues/3989

## Proposed Changes

* Add a generic component for Themeshowcase that displays the themes in a swiper carousel
* Add two components for Premium and Partner themes
* The implementation is behind a flag and not visible to users

## Testing Instructions
- Use the Calypso live link and go to /themes page in incognito.
- Append the feature flag `flags=themes/discovery-lots`
- Notice that we display three carousels:
- - Premium themes 
- - Partner themes 
- Ignore the unstyled carousel, which will be handled in a separate PR.

Screenshot 2023-10-06 at 17 27 40

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/components/theme-collection/index.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import './style.scss';
+import Swiper from 'swiper';
+import { Navigation, Keyboard, Mousewheel } from 'swiper/modules';
+import { ThemeBlock } from 'calypso/components/themes-list';
+import 'swiper/css';
+import 'swiper/css/navigation';
+import 'swiper/css/keyboard';
+import 'swiper/css/mousewheel';
+
+interface ThemeCollectionProps {
```
Comment (BogdanUngureanu) [inline client/components/theme-collection/index.tsx:7]: This interface needs to be refactored. I've duplicated the props in the PremiumCollection for now.

We'll need a generic interface that this interface will extend. 

We can do it in a separate PR.

File: client/components/theme-collection/index.tsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+			}
+		}
+
+		return () => {
+			swiperInstance.current?.destroy();
+		};
+	}, [ themes ] );
+
+	if ( themes === null ) {
+		return <></>; // @todo Add LoadingPlaceholders instead.
```
Comment (BogdanUngureanu) [inline client/components/theme-collection/index.tsx]: We should mark themes as nullable array. When themes is null, we should display a LoadingPlaceholder to avoid a flickering screen.

File: client/components/theme-collection/index.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	isActive,
+	getPrice,
+	isInstalling,
+}: ThemeCollectionProps ): ReactElement {
+	const swiperInstance = useRef< SwiperType | null >( null );
+	useEffect( () => {
+		if ( themes !== null && themes.length > 0 ) {
+			const el = document.querySelector( '.swiper-container' ) as HTMLElement;
+			if ( el ) {
+				Promise.all( [
```
Comment (BogdanUngureanu) [inline client/components/theme-collection/index.tsx:32]: We have to import the swiper package dynamically because it doesn't offer a CommonJS version. Because of this, the SSR build will fail.

From a performance standpoint, I don't think it will cause any issues; we are already loading the ThemeCards dynamically anyway.

Comment (dsas) [reply client/components/theme-collection/index.tsx:32]: This should be a code comment really

Comment (BogdanUngureanu) [reply client/components/theme-collection/index.tsx:32]: Good point!

File: client/components/theme-collection/index.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+						<Icon icon={ chevronLeft } />
+					</Button>
+					<Button className="theme-collection__carousel-nav-button theme-collection__carousel-nav-button--next">
+						<Icon icon={ chevronRight } />
+					</Button>
+				</div>
+				<div className="theme-collection__list-wrapper swiper-wrapper">
+					{ themes.map( ( theme, index ) => (
+						<div
+							key={ `theme-collection-container-${ collectionSlug }-${ index }` }
```
Comment (dsas) [inline client/components/theme-collection/index.tsx]: `index` is just the array index, react keys should use constant indexes not numeric array indexes because the zeroth item on the first render, could be the first item on the second render and it would update the wrong element.

This is probably only going to be a problem in this case if someone nulls out the react state and then someone changes curation order and so it refetches things, or something.

But still, easy enough to use theme.slug or something to generate a key.

Comment (BogdanUngureanu) [reply client/components/theme-collection/index.tsx]: Sure, I think we can change it. The numeric approach was taken from the existing implementation of the theme list component.

File: client/my-sites/themes/themes-selection.jsx [hunk 5]
```diff
@@ -265,10 +269,51 @@ class ThemesSelection extends Component {
 
 		const interlacedThemes = interlaceThemes( themes, wpOrgThemes, query.search, isLastPage );
 
+		const isActiveInLits = config.isEnabled( 'themes/discovery-lits' ) && this.props.isLoggedIn;
+		const isActiveInLots = config.isEnabled( 'themes/discovery-lots' ) && ! this.props.isLoggedIn;
```
Comment (Copons) [inline client/my-sites/themes/themes-selection.jsx]: Nitpick: I find that `isActiveIn...` does not describe what we are doing here, and IMHO it's a bit confusing with all the `lits` and `lots`. 🙂 

How about something like this:

```js
import { isEnabled } from '@automattic/calypso-config';

// ...

const {
  // ...
  isLoggedIn,
} = this.props;

// Consider other names like `isThemeCollectionEnabled` etc.
const isDiscoveryEnabled =
  ( isEnabled( 'themes/discovery-lits' && isLoggedIn ) ||
  ( isEnabled( 'themes/discovery-lots' && ! isLoggedIn );


```

File: client/components/theme-collection/index.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	const swiperContainerId = `swiper-container-${ collectionSlug }`;
+
+	useEffect( () => {
+		if ( themes !== null && themes.length > 0 ) {
+			const el = document.querySelector( `#${ swiperContainerId }` ) as HTMLElement;
+			if ( el ) {
+				/**
+				 * We have to import the swiper package dynamically because it doesn't offer a CommonJS version. Because of this, the SSR build will fail.
+				 *
+				 * From a performance standpoint, I don't think it will cause any issues; we are already loading the ThemeCards dynamically anyway.
```
Comment (Copons) [inline client/components/theme-collection/index.tsx]:
```suggestion
				 * From a performance standpoint, it should not cause any issues; we are already loading the ThemeCards dynamically anyway.
```

File: client/components/theme-collection/index.tsx [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+	getButtonOptions,
+	getActionLabel,
+	isActive,
+	getPrice,
+	isInstalling,
+}: ThemeCollectionProps ): ReactElement {
+	const swiperInstance = useRef< SwiperType | null >( null );
+	const swiperContainerId = `swiper-container-${ collectionSlug }`;
+
+	useEffect( () => {
```
Comment (Copons) [inline client/components/theme-collection/index.tsx:23]: This is initializing Swiper at every render of the component.

File: client/components/theme-collection/create-map-state-to-props.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+		const siteId = getSelectedSiteId( state ) as unknown as string;
+
+		const themes = ( getThemesForQueryIgnoringPage( state, 'wpcom', query ) as never[] ) ?? []; // We should change it to a nullable value so that we can use LoadingPlaceholders.
+		return {
+			...ownProps,
+			themes,
+			siteId,
+		};
+	};
+};
```
Comment (Copons) [inline client/components/theme-collection/create-map-state-to-props.ts]: Let's use hooks instead of `connect` (which is soft-deprecated) for selecting from the Redux state.

e.g.
```js
/* use-theme-collection.ts */

import { useSelector } from 'calypso/state';

export function useThemeCollection( query ) {
  const siteId = useSelector( getSelectedSiteId );
  const themes = useSelector(
    ( state ) => getThemesForQueryIgnoringPage( state, 'wpcom', query ) ?? []
  );

  return { siteId, themes };
}


/* premium-themes-collection.tsx */

import useThemeCollection from './use-theme-collection';

const query = {
  // ...
};

export default function PremiumThemesCollection( props ) {
  const { siteId, themes } = useThemeCollection( query );

  // ...
}
```

File: client/components/theme-collection/collection-types/partner-themes-collection.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+				collectionSlug="partner-themes"
+			/>
+		</>
+	);
+}
+
+export default connect(
+	createWpcomThemeCollectionMapStateToProps( query ),
+	null
+)( localize( PartnerThemesCollection ) );
```
Comment (Copons) [inline client/components/theme-collection/collection-types/partner-themes-collection.tsx]: Besides replacing `connect` with a hook (as mentioned in another comment), we should also localize via hook!

```js
import { useTranslate } from 'i18n-calypso';

export default function FoobarCollection() {
  const translate = useTranslate();

  return <>{ translate( 'Foobar!' ) }</>;
}
```

⚠️ All these notes about Redux, `connect`, and hooks are relevant for all the collection-types.

File: client/components/theme-collection/create-map-state-to-props.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * Create a mapStateToProps closure for theme collections based on a theme query object.
+ * @param query
+ */
+export const createWpcomThemeCollectionMapStateToProps = ( query: object ) => {
+	return ( state: IAppState, ownProps: ThemesCollectionProps ) => {
+		const siteId = getSelectedSiteId( state ) as unknown as string;
+
+		const themes = ( getThemesForQueryIgnoringPage( state, 'wpcom', query ) as never[] ) ?? []; // We should change it to a nullable value so that we can use LoadingPlaceholders.
```
Comment (Copons) [inline client/components/theme-collection/create-map-state-to-props.ts]: I have no idea what `never[]` is, but it sounds terrifying.

File: client/components/theme-collection/index.tsx [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+	subheading,
+	themes,
+	getScreenshotUrl,
+	siteId,
+	getButtonOptions,
+	getActionLabel,
+	isActive,
+	getPrice,
+	isInstalling,
+}: ThemeCollectionProps ): ReactElement {
```
Comment (Copons) [inline client/components/theme-collection/index.tsx]: Nitpick: let's try to keep props ordered alphabetically (in all component functions and calls).

File: client/components/theme-collection/index.tsx [hunk 12]
```diff
@@ -0,0 +1,133 @@
+import { Button } from '@wordpress/components';
+import { Icon, chevronLeft, chevronRight } from '@wordpress/icons';
+import { ReactElement, useEffect, useRef } from 'react';
+import './style.scss';
+import { Swiper as SwiperType } from 'swiper/types';
```
Comment (Copons) [inline client/components/theme-collection/index.tsx:5]: There is a [Swiper React](https://swiperjs.com/react) set of components, which I can see in the Calypso node_modules (`wp-calypso/node_modules/swiper/swiper-react.mjs`).

What's the reason for using the vanilla JS version instead?

**Review by Copons — CHANGES_REQUESTED** [review]
I think we are on the right direction.

I've noted a few changes I'd like to see (mostly around using modern React), and an open question about using Swiper React.

File: client/components/theme-collection/index.tsx (hunk 12 above)
Comment (dsas) [reply client/components/theme-collection/index.tsx:5]: We did use those originally! The swiper people [say they're probably going to remove them](https://swiperjs.com/react) and recommend migrating to the Swiper Element API. The GitHub issues had people struggling with various uses of the Swiper Element API so I switched to ye olde vanilla js.

File: client/my-sites/themes/themes-selection.jsx (hunk 5 above)
Comment (BogdanUngureanu) [reply client/my-sites/themes/themes-selection.jsx]: Sure! :)

File: client/components/theme-collection/index.tsx (hunk 12 above)
Comment (Copons) [reply client/components/theme-collection/index.tsx:5]: I guess not a biggie, although I liked the React-y version.

Their warning is super vague ("likely", "in the future", etc.), and I wonder whether we would be safe with it for the time being. 🤔 

Oh well, if this works, then never mind my personal preferences. 🙂

File: client/components/theme-collection/index.tsx (hunk 7 above)
Comment (BogdanUngureanu) [reply client/components/theme-collection/index.tsx:23]: Wouldn't the Swiper initialize just once because we pass `themes` const and because we check it for null values? 🤔

File: client/components/theme-collection/index.tsx (hunk 12 above)
Comment (dsas) [reply client/components/theme-collection/index.tsx:5]: I liked it too, and I may just have the wrong instincts given our approach to dependency updates etc.

Using the react version seems like storing up a problem for the future. Swiper seems to move faster than we do in terms of only supporting the latest stuff and dropping support for older stuff. The last thing we want is more work when we're reacting to a security problem in 2028 when we're jumping from 10.1 to 20.1

File: client/components/theme-collection/index.tsx [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+}: PropsWithChildren< { collectionSlug: string; themeId: string } > ) {
+	return (
+		<div
+			key={ `theme-collection-container-${ collectionSlug }-${ themeId }` }
+			className="theme--collection__list-item swiper-slide"
+		>
+			{ children }
+		</div>
+	);
+}
```
Comment (Copons) [inline client/components/theme-collection/index.tsx]: What's the reasoning for `CollectionListItem`?

For what I can see, the only significant bit is the `swiper-slide` CSS class needed by Swiper.
Everything else seems superfluous, and even `swiper-slide` technically is. 😄 

If you need to style a slide (e.g. to give them a width), you can do that with `.theme-collection__list-wrapper > *`.
It then becomes the responsibility of the consumer to ensure the ThemeCollection children have an appropriate/stylable wrapper div.

If you need a specific slide class for the Swiper initialization, why not let the consumer define it?

e.g.
```jsx
// consumer
<ThemeCollection slideClass="foobar__slide" />

// Collection
function ThemeCollection( { slideClass } ) {
  // ...
  new Swiper( el, {
    // all the settings
    slideClass
  } );
  // ...
}
```

This would also give us the opportunity to offer further slider customizations to various Collection consumers (not the case right now, but it's good to keep in mind for the future).

File: client/components/theme-collection/index.tsx [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+						<Icon icon={ chevronRight } />
+					</Button>
+				</div>
+				<div className="theme-collection__list-wrapper swiper-wrapper">{ children }</div>
+			</div>
+		</div>
+	);
+}
+
+export function CollectionListItem( {
```
Comment (Copons) [inline client/components/theme-collection/index.tsx]: If we want to keep this, let's:
- Rename it to `ThemeCollectionItem`.
- Move it to its own separate `theme-collection-item` file.

File: client/components/theme-collection/index.tsx [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+	collectionSlug,
+	themeId,
+	children,
+}: PropsWithChildren< { collectionSlug: string; themeId: string } > ) {
+	return (
+		<div
+			key={ `theme-collection-container-${ collectionSlug }-${ themeId }` }
```
Comment (Copons) [inline client/components/theme-collection/index.tsx]: `key` is only needed when the element is a child of a list.

It is not the case here; it is when the `Item` component is used (like you do in the two collection type components).

File: client/components/theme-collection/index.tsx (hunk 13 above)
Comment (Copons) [reply client/components/theme-collection/index.tsx]: As a counterpoint, I think there's value in offering a `ThemeCollectionItem` component, which might decrease the consumer's cognitive load.

I'd recommend keeping it as simple as possible, with only the essential bits for now and maybe extending it in the future if needed. E.g.

```jsx
export default function ThemeCollectionItem( { children } ) {
  return (
    <div className="theme-collection__item swiper-slide">
      { children }
    </div>
  );
}
```

File: client/components/theme-collection/index.tsx [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+			className="theme--collection__list-item swiper-slide"
```
Comment (Copons) [inline client/components/theme-collection/index.tsx]:
```suggestion
			className="theme-collection__list-item swiper-slide"
```

(see [CSS guidelines](https://github.com/Automattic/wp-calypso/blob/7752207da247005f4b24898bad1807e16a474d92/docs/coding-guidelines/css.md))

File: client/components/theme-collection/collection-types/partner-themes-collection.tsx [hunk 17]
```diff
@@ -0,0 +1,49 @@
+import { useTranslate } from 'i18n-calypso';
+import { memo, ReactElement } from 'react';
+import * as React from 'react';
```
Comment (Copons) [inline client/components/theme-collection/collection-types/partner-themes-collection.tsx]: We shouldn't need this.

File: client/components/theme-collection/collection-types/premium-themes-collection.tsx [hunk 18]
```diff
@@ -0,0 +1,48 @@
+import { useTranslate } from 'i18n-calypso';
+import { memo, ReactElement } from 'react';
+import * as React from 'react';
```
Comment (Copons) [inline client/components/theme-collection/collection-types/premium-themes-collection.tsx]: We shouldn't need this.

File: client/components/theme-collection/themes-collection-props.ts [hunk 19]
```diff
@@ -0,0 +1,8 @@
+export interface ThemesCollectionProps {
+	getActionLabel: () => string;
+	getButtonOptions: () => void;
+	getPrice: () => string;
+	getScreenshotUrl: ( themeId: string ) => string;
+	isActive: () => boolean;
+	isInstalling: () => boolean;
+}
```
Comment (Copons) [inline client/components/theme-collection/themes-collection-props.ts]: I think this is not needed anymore.

File: client/components/theme-collection/index.tsx [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+interface ThemeCollectionProps {
+	collectionSlug: string;
+	heading: string;
+	subheading: ReactElement;
+}
+
+export default function ThemeCollection( {
+	collectionSlug,
+	heading,
+	subheading,
```
Comment (Copons) [inline client/components/theme-collection/index.tsx]: How do you feel about renaming these into `title` and `description`? IMHO it's more clear at first glance.

File: client/components/theme-collection/use-theme-collection.ts [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+import { getThemesForQueryIgnoringPage } from 'calypso/state/themes/selectors';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+interface ThemesQuery {
+	page: number;
+	number: number;
+	tier: string;
+	filter: string;
+	search: string;
+	collection: string;
```
Comment (Copons) [inline client/components/theme-collection/use-theme-collection.ts]: Is `collection` actually a query property?

File: client/components/theme-collection/index.tsx [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+	children,
+}: PropsWithChildren< ThemeCollectionProps > ): ReactElement {
+	const swiperInstance = useRef< SwiperType | null >( null );
+	const swiperContainerId = `swiper-container-${ collectionSlug }`;
+
+	const [ isSwiperLoaded, setSwiperLoaded ] = useState( false );
+	useEffect( () => {
+		if ( ! isSwiperLoaded ) {
```
Comment (dsas) [inline client/components/theme-collection/index.tsx]: I think this can get this component and swiper out of sync. If the swiper instance gets cleaned up (for whatever reason) then `isSwiperLoaded` will remain true and the swiper instance won't get recreated.

```suggestion
		if ( ! isSwiperLoaded || ! swiperInstance?.current || swiperInstance.current?.destroyed ) {
```

File: client/my-sites/themes/themes-selection.jsx [hunk 23]
```diff
@@ -247,6 +250,43 @@ class ThemesSelection extends Component {
...
+	};
+
+	isInstallingTheme = ( theme ) => {
+		return this.props.isInstallingTheme( theme );
+	};
+
+	getCarouselBlock = ( collectionSlug, theme, index ) => {
+		return (
+			<ThemeBlock
+				collectionSlug={ collectionSlug }
```
Comment (Copons) [inline client/my-sites/themes/themes-selection.jsx]: I don't think we need this.

File: client/my-sites/themes/themes-selection.jsx [hunk 24]
```diff
@@ -247,6 +250,43 @@ class ThemesSelection extends Component {
...
+				getActionLabel={ this.getActionLabel }
+				getButtonOptions={ this.getOptions }
+				getPrice={ this.getPremiumThemePrice }
+				getScreenshotUrl={ this.getScreenshotUrl }
+				index={ index }
+				isActive={ this.isThemeActive }
```
Comment (Copons) [inline client/my-sites/themes/themes-selection.jsx]: Eventually, we will not show the active theme in both collections and regular lists.

We can filter out the active theme from `themes` and remove this prop from the collection's items.

File: client/my-sites/themes/themes-selection.jsx [hunk 25]
```diff
@@ -247,6 +250,43 @@ class ThemesSelection extends Component {
...
+		return this.props.getScreenshotUrl( theme, themeOptions );
+	};
+
+	isThemeActive = ( theme ) => {
+		return this.props.isThemeActive( theme );
+	};
+
+	isInstallingTheme = ( theme ) => {
+		return this.props.isInstallingTheme( theme );
+	};
```
Comment (Copons) [inline client/my-sites/themes/themes-selection.jsx]: I'm missing something here.

Why are we wrapping these functions instead of passing the props through?

E.g. can't we do this?
```jsx
<ThemeBlock
  getActionLabel={ this.props.getActionLabel( theme ) }
</ThemeBlock>
```

File: client/my-sites/themes/themes-selection.jsx [hunk 26]
```diff
@@ -261,14 +301,27 @@ class ThemesSelection extends Component {
...
+
 		return (
 			<div className="themes__selection">
 				<QueryThemes query={ query } siteId={ source } />
 				{ shouldFetchWpOrgThemes && <QueryThemes query={ wpOrgQuery } siteId="wporg" /> }
+
+				{ isDiscoveryEnabled && (
+					<>
+						<PremiumThemesCollection>{ this.getCarouselBlock }</PremiumThemesCollection>
+						<PartnerThemesCollection>{ this.getCarouselBlock }</PartnerThemesCollection>
```
Comment (Copons) [inline client/my-sites/themes/themes-selection.jsx]: I think this is super confusing and overcomplicated.

Throwing an idea into the mix: collection-specific **information** instead of collection-specific **components**.

Hopefully the example could clarify what I'm thinking:

_(Note: this is all untested!)_
```jsx
// theme-collection/constants.jsx

export const THEME_COLLECTIONS = {
  premium: {
    query: { /* ... */ },
    heading: translate( 'Premium Themes' ),
    subheading: <p>{ translate( 'Lorem ipsum dolor sit amet' ) }</p>,
  }
}
```

```jsx
// themes-selection.jsx

getThemeCollection = ( slug ) => {
  const {
    query,
    heading,
    subheading,
  } = THEME_COLLECTIONS[ slug ];
  
  const { themes } = useThemeCollection( query );

  return (
    <>
      <QueryThemes query={ query } siteId="wpcom" />
      <ThemeCollection
        slug={ slug }
        heading={ heading }
        subheading={ subheading }
      >
        { themes && themes.map( ( theme, index ) => (
          <ThemeCollectionItem key={ theme.id }>
            <ThemeBlock /* all the props */ />
          </ThemeCollectionItem>
        ) ) }
      </ThemeCollection>
    </>
  );
}

render() {
  return(
    <>
      { isDiscoveryEnabled &&
        getThemeCollection( 'premium' )
      }
    </>
  );
}
```

Off the top of my head, this has 2 big advantages:
- Simplifies the hell out of everything, reducing the prop drilling.
- Does not couple any theme fetching logic with collections.
  - By and large, we're going to show this in 1 or 2 places (TS and DP). I haven't checked, but I presume that DP has a completely different theme fetching logic (IIRC it uses a different endpoint even).
  - If we create a collection-specific component using `QueryThemes`, we are fundamentally limiting it to the TS, which would remove its usefulness altogether (what's the point of such a specific component if we are only going to use it once?).

File: client/my-sites/themes/themes-selection.jsx (hunk 25 above)
Comment (BogdanUngureanu) [reply client/my-sites/themes/themes-selection.jsx]: I had it like that but it triggers a re-render when the props from ThemeSelection are called.

File: client/components/theme-collection/collection-types/premium-themes-collection.tsx [hunk 27]
```diff
@@ -32,13 +33,13 @@ function PremiumThemesCollection( {
 			>
 				{ themes &&
 					themes.map( ( theme, index ) => (
-						<CollectionListItem
+						<ThemeCollectionItem
 							key={ theme.id }
 							collectionSlug="premium-themes"
```
Comment (dsas) [inline client/components/theme-collection/collection-types/premium-themes-collection.tsx]: `premium-themes` could be moved up into a const rather than repeated.

File: client/components/theme-collection/use-theme-collection.ts (hunk 21 above)
Comment (BogdanUngureanu) [reply client/components/theme-collection/use-theme-collection.ts]: yup

File: client/components/theme-collection/index.tsx (hunk 20 above)
Comment (BogdanUngureanu) [reply client/components/theme-collection/index.tsx]: sure :)

File: client/components/theme-collection/themes-collection-props.ts (hunk 19 above)
Comment (BogdanUngureanu) [reply client/components/theme-collection/themes-collection-props.ts]: No, I don't think so. :)

File: client/my-sites/themes/themes-selection.jsx (hunk 26 above)
Comment (BogdanUngureanu) [reply client/my-sites/themes/themes-selection.jsx]: > THEME_COLLECTIONS.premium.query

I have some mixed feelings about keeping logic code in a constants file, separated from the actual code. I'm fine with keeping the texts (e.g. heading), but not sure about the query.

It's interesting. However, there's one problem with this approach. The code would add up and will increase even more the complexity of the ThemeSelection.

An alternative is to have a new component called `ThemeCollectionsLayout` that would:
- define the collections (query, headings, etc).
- construct the layout (collections, upgrade banners)

Keep in mind that besides the collection components, we still need to add the routing logic. It will add up. :)

Comment (dsas) [reply client/my-sites/themes/themes-selection.jsx]: > in a constants file

You could just call the file "collection-definitions.ts" if that helps

Comment (Copons) [reply client/my-sites/themes/themes-selection.jsx]: > I have some mixed feelings about keeping logic code in a constants file

That really isn't logic though. It's the options argument for the themes endpoint. It would be the same as adding, like, `themesPerPage: 10`.

File: client/my-sites/themes/theme-collections-layout.tsx [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+		collectionSlug,
+		title,
+		description,
+		query,
+	}: {
+		collectionSlug: string;
+		title: string;
+		description: ReactElement;
+		query: ThemesQuery;
+	} ): ReactElement => {
```
Comment (Copons) [inline client/my-sites/themes/theme-collections-layout.tsx]: Let's turn these types into a `ShowcaseThemeCollectionProps` interface for the sake of my eyes. 😆

File: client/my-sites/themes/theme-collections-layout.tsx [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+						) ) }
+				</ThemeCollection>
+			</>
+		);
+	};
+
+	return (
+		<>
+			<ShowcaseThemeCollection
+				key={ THEME_COLLECTIONS.premium.collectionSlug }
```
Comment (Copons) [inline client/my-sites/themes/theme-collections-layout.tsx]: We don't need keys here, it's not a list, it's just two instances of the same component.

File: client/my-sites/themes/theme-collections-layout.tsx [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+	getOptions: ( themeId: string ) => void;
+	getScreenshotUrl: ( themeId: string ) => string;
+}
+
+function ThemeCollectionsLayout( {
+	getActionLabel,
+	getOptions,
+	getScreenshotUrl,
+}: ThemesCollectionsLayoutProps ) {
+	const ShowcaseThemeCollection = ( {
```
Comment (Copons) [inline client/my-sites/themes/theme-collections-layout.tsx]: I'm not convinced we should declare a component inside a component.
I presume the intent is to remove one level of props drilling, but it then becomes an antipattern and performance concern.
If we want to streamline this, we could do something like this instead:

```jsx
function ShowcaseThemeCollection( {
	collectionSlug,
	title,
	description,
	query,
	getActionLabel,
	getOptions,
	getScreenshotUrl,
} ) {
	// ...
}

function ThemeCollectionsLayout( props ) {
	return (
		<>
			<ShowcaseThemeCollection
				{ ...THEME_COLLECTIONS.premium }
				{ ...props }
			/>
		</>
	);
}
```

File: client/components/theme-collection/index.tsx [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+
+		return () => {
+			swiperInstance.current?.destroy();
+		};
+	}, [ swiperContainerId, isSwiperLoaded ] );
+
+	return (
+		<div className="theme-collection__container ">
+			<h2>{ title }</h2>
+			{ description }
```
Comment (Copons) [inline client/components/theme-collection/index.tsx]: I'd recommend adding some specificity. I'm sure it will turn out to be useful. 🙂
```suggestion
			<h2 className="theme-collection__title">{ title }</h2>
			<div className="theme-collection__description">{ description }<div>
```

File: client/my-sites/themes/theme-collections-layout.tsx [hunk 32]
```diff
@@ -0,0 +1,76 @@
+import { memo, ReactElement } from 'react';
+import QueryThemes from 'calypso/components/data/query-themes';
+import ThemeCollection from 'calypso/components/theme-collection';
+import ThemeCollectionItem from 'calypso/components/theme-collection/theme-collection-item';
+import { ThemeBlock } from 'calypso/components/themes-list';
+import { THEME_COLLECTIONS } from 'calypso/my-sites/themes/collection-definitions';
+import { ThemesQuery, useThemeCollection } from 'calypso/my-sites/themes/use-theme-collection';
+
+interface ThemesCollectionsLayoutProps {
```
Comment (Copons) [inline client/my-sites/themes/theme-collections-layout.tsx]:
```suggestion
interface ThemeCollectionsLayoutProps {
```

File: client/my-sites/themes/collection-definitions.tsx [hunk 33]
```diff
@@ -0,0 +1,30 @@
...
+			collection: 'recommended',
+			filter: '',
+			number: 10,
+			page: 1,
+			search: '',
+			tier: 'premium',
+		},
+		title: translate( 'Premium Themes' ),
+		collectionSlug: 'premium-themes',
+		description: <p>{ translate( 'Lorem ipsum dolor sit amet' ) }</p>,
```
Comment (Copons) [inline client/my-sites/themes/collection-definitions.tsx]: Let's not translate placeholder text please!

File: client/components/theme-collection/index.tsx (hunk 31 above)
Comment (dsas) [reply client/components/theme-collection/index.tsx]: 100% will, I already did the same in https://github.com/Automattic/wp-calypso/pull/82519/files

File: client/my-sites/themes/theme-showcase.jsx [hunk 34]
```diff
@@ -402,26 +403,62 @@ class ThemeShowcase extends Component {
...
 		const tabKey = this.getSelectedTabFilter().key;
+
+		const isDiscoveryEnabled =
+			( config.isEnabled( 'themes/discovery-lits' ) && this.props.isLoggedIn ) ||
+			( config.isEnabled( 'themes/discovery-lots' ) && ! this.props.isLoggedIn );
+
 		switch ( tabKey ) {
 			case staticFilters.MYTHEMES?.key:
 				return <ThemesSelection { ...themeProps } />;
+			case staticFilters.RECOMMENDED.key:
```
Comment (dsas) [inline client/my-sites/themes/theme-showcase.jsx:414]: At some point, maybe when we drop the feature flag, maybe we should think about renaming RECOMMENDED. Don't think we need to change now though.

**Comment by dsas** [comment]
Tracks check:
 * `calypso_upgrade_nudge_impression` (with `{ cta_name : theme-upsell }` )  is fired for each theme
   * Including those that are off-screen due to the carousel. Given live lots triggers the same event for themes that are off-screen due to scroll, maybe this is fine.
 * Scrolling the carousel whether by button, mousewheel etc doesn't trigger any tracks events, even when reaching the end. Live lots doesn't trigger an equivalent event
 * ❗Clicking on a theme in the carousel doesn't trigger a `calypso_themeshowcase_theme_click` event, unlike live lots

I don't think this needs changing in this PR, we already have a task later in the proejct for checking and creating tracks events, but thought it was worth being aware.

**Review by Copons — APPROVED** [review]
💥 LGTM 💥 

Thank you Bogdan for this PR's rollercoaster!

@dsas would you mind creating a project issue for the Tracks event you noted in the comment above? https://github.com/Automattic/wp-calypso/pull/82254#issuecomment-1750203937

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/9255148

Thank you @BogdanUngureanu for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/9255148).
